stor/s3.py

```diff
@@ -765,3 +769,34 @@ def to_url(self):
         """Returns HTTP url for object (virtual host-style)"""
         return u'https://{bucket}.s3.amazonaws.com/{key}'.format(bucket=self.bucket,
                                                                  key=self.resource)
+
+    def restore(self, tier='Bulk', days=10):
+        """Issue a restore command for a single object from glacier.
+
+        Args:
+            tier (str, optional): restore speed (see Glacier docs for details)
+            days (int, optional): number of days to keep data in S3 post-restore.
+        Returns:
+            dict or None: restore response from S3 client
+        Note:
+            Calling ``restore()`` on a directory will not work correctly. Only use this for single
+            objects! (you can always do ``[s.restore() for s in stor.list(<directory>)]``)
+
+        Ignores RestoreAlreadyInProgressError and AlreadyRestoredError (note that you can't force
```

Comment:
do we actually want to catch these errors?  What's wrong with raising to user?

Reply:
RestoreAlreadyInProgress happens when you do this:

```
path = stor.Path('s3://bucket/key/obj')
path.restore()
path.restore() # at this point Amazon returns a 409 Conflict, RestoreAlreadyInProgress error
```

AlreadyRestoredError:

```
path = stor.Path('s3://bucket/key/obj')
assert path.stat()['StorageClass'] != 'GLACIER'
path.restore()
```

in both cases, the user's intention (ensure this object is restored from Glacier to S3) is already satisfied (either because restore has already started or because it already is in S3).  In my view, bubbling up the error just makes the API more complicated to understand. (whereas without that error, you can just iterate over all objects on a bucket, hit restore on all of them and not need to know which errors to catch)

Reply:
maybe make an optional kwarg to raise or suppress?

Reply:
agree default should be to suppress. logger is good too.

Reply:
if we need an optional kwarg at some future time we can add it. there's already logging at debug level